### ISSUE TITLE
[netcore][noise] Add timestamp challenge to first Noise handshake message

### ISSUE THREAD

**Issue by phlip9** [libra/libra#1822]
## Motivation

As a part of the DoS/QoS roadmap, we want to minimize work done on invalid/malicious inbound communication. Part of this involves rejecting Noise handshakes from unauthorized peers with as little work and communication as possible, where "unauthorized peers" means non-validators trying to connect to a validator on a port where the validator is only accepting connections from other validators.

Following #1400, we will use the Noise_IK handshake for establishing secure connections. For reference, the Noise_IK handshake looks like:

```
# Noise_IK handshake
> e, es, s, ss                          # if s not in validator_set { disconnect }
< e, ee, se
```

We are concerned with three DoS attack scenarios:

### [1] Attacker sends random handshake bytes

To reject, we need to do one DHKE to read `s` and recognize the unauthorized `s` key.

### [2] Attacker sends handshake with actual validator's static pubkey but fails to generate valid MAC

To reject, we need to do two DHKE: one to read `s` and one to derive the shared secret from `ss` for the MAC tag. Since the attacker doesn't know the actual validator's network private key, they fail to derive the `ss` shared secret and we can reject after reading the invalid MAC tag.

### [3] Passive eavesdropper records an actual validator-to-validator handshake and replays these handshake messages

In this case, the attacker has observed a valid first-handshake message `> e, es, s, ss` between two actual validators. Without any specific protections, the attacker can replay this message as many times as they wish to consume available inbound connection slots, replace an existing and honest connection between the two validators, and hold open this connection until the transport upgrade times out (currently 30s). The reason for this is that the listener cannot actually validate this connection until the attacker sends their first message after the handshake, which will fail to present a valid MAC tag. Even worse, the attacker can keep replaying this handshake message as it is always valid.

## Proposal

To combat [3], we can add a "proof of recency" to the initial Noise handshake message. To prove that the message is recent, we can frame the first message with a timestamp. The timestamp can then be used in the [Noise handshake prologue](http://www.noiseprotocol.org/noise.html#prologue), which will tie this session to the timestamp. When accepting new inbound connections, we can then reject any inbound handshake when `handshake_time <= previous_handshake_time`.

If an attacker then tries to replay a handshake, then it will be rejected as `handshake_time <= previous_handshake_time`. This also means that they can't replace any existing connections. Even if the listener restarts, the attacker can at most hold open this one connection for ~30s until the timeout hits. If the attacker tries to spoof a different timestamp, then `s` will fail to decrypt properly and we can reject the random `s` we get.

We should probably also ensure that we read `handshake_time` from a monotonic clock.

Note: I thought we might need to add another reject condition `handshake_time <= local_time + 30sec` or something, but I think the current `handshake_time <= previous_handshake_time` condition covers all the cases.

cc @bothra90 @mimoo @aching

**Comment by mimoo** [comment]
https://github.com/libra/libra/pull/3951 does this with the current parameters:

```

/// we're willing to tolerate a client timestamp 10 seconds in the future
const MAX_FUTURE_TIMESTAMP: u64 = 10;

/// we store client timestamps (to prevent replay) for up to 60 seconds
const EXPIRATION_TIMESTAMP: u64 = 60;
```

**Comment by phlip9** [comment]
Closed by #3951
